### ISSUE TITLE
Instability in `test_pull_timeline`

### ISSUE THREAD

**Issue by jcsp** [neondatabase/neon#9731]
https://neon-github-public-dev.s3.amazonaws.com/reports/pr-9646/11788052818/index.html#testresult/5a4b2cac9be55bb0/

```
command failed: error sending request for url (http://127.0.0.1:17120/configure)
```

**Comment by arpad-m** [comment]
I've went over the seven runs where it failed in the last 14 days, and they all failed on the same reconfiguration request. Log excerpt:

```
2024-11-08 20:03:27.644 INFO [test_wal_acceptor.py:1848] reconfiguing compute with new config to verify that it works
2024-11-08 20:03:27.644 INFO [neon_cli.py:72] Running command "/tmp/neon/bin/neon_local endpoint reconfigure ep-1 --tenant-id 0b0b1497dff3937d1002c5fbbb58e6ab --safekeepers 1,3,4"
2024-11-08 20:03:57.793 INFO [neon_cli.py:137] Run ['/tmp/neon/bin/neon_local', 'endpoint', 'reconfigure', 'ep-1', '--tenant-id', '0b0b1497dff3937d1002c5fbbb58e6ab', '--safekeepers', '1,3,4'] failed:
  stdout:
    Getting claims for path debug/v1/tenant/0b0b1497dff3937d1002c5fbbb58e6ab/locate
    Got claims Claims { tenant_id: None, scope: Admin } for path debug/v1/tenant/0b0b1497dff3937d1002c5fbbb58e6ab/locate
  stderr:
    command failed: error sending request for url (http://127.0.0.1:17035/configure)

    Caused by:
        operation timed out
```

All 7 reproductions were on postgres 14 or 15, and all on release.

**Comment by arpad-m** [comment]
Looking at the compute logs, it seems that the configure operation takes 30 seconds:

```
2024-11-08T20:03:27.660149Z  INFO http request{otel.name=/configure http.method=POST}: serving /configure POST request
...
2024-11-08T20:03:57.780342Z  INFO configurator_main_loop: compute node configured
```

Looking at four other reproductions of the seven, it's always 30 seconds. Maybe some timeout that it's running into?

When it works, the operation is over much more quickly, within a second, say [here](https://neon-github-public-dev.s3.amazonaws.com/reports/main/11748357494/index.html#testresult/edec47bbb2cfd3e8/):

```
2024-11-08T20:04:08.061696Z  INFO http request{otel.name=/configure http.method=POST}: serving /configure POST request
...
2024-11-08T20:04:09.172912Z  INFO configurator_main_loop: compute node configured
```

The compute logs for the buggy case show this around the "gap" of 30 seconds. Apparently it tries to upload traces:
```
PG:2024-11-08 20:03:27.669 GMT [351693] LOG:  [NEON_SMGR] [shard 0] libpagestore: connected to 'postgresql://no_user@localhost:17020' with protocol version 2
OpenTelemetry trace error occurred. error sending request for url (http://localhost:4318/v1/traces)
PG:2024-11-08 20:03:57.693 GMT [367892] LOG:  [WP] connecting with node 127.0.0.1:17023
```

This OpenTelemetry trace error can either be a coincidence, or actually be due to us.

From the logs, it seems to hang in the second `self.pg_reload_conf()?;` invocation in `ComputeNode::reconfigure`. This function invokes the `pg_ctl` binary, and then waits for it to terminate.
We throw away the stdout/stderr from that binary, and we also don't print the stdout/stderr, which might be useful here.

**Comment by arpad-m** [comment]
Hmm since today 11:05 UTC, it seems to be way more flaky:

![Image](https://github.com/user-attachments/assets/1a872dce-a0a8-4039-b203-7b80557c65a6)

If I look [at the main branch commits](https://github.com/neondatabase/neon/commits/main/?since=2024-11-20&until=2024-11-20), I see a PR having been merged on 11:07 UTC which is safekeeper related: https://github.com/neondatabase/neon/pull/9364 .

But the occurence from 11:05 (a8c1f676798cafce738e8c446d8d5a77d2380f6a) didn't include it, nor did the second one on 11:08 (e5024a5c05016c30dec7897aca22d1040a340f63). Both include https://github.com/neondatabase/neon/pull/9717 however.

**Comment by arpad-m** [comment]
I see 23 flaky failures and 364 total runs of the test, but maybe only a third of the tests actually is susceptible to the issue. So one can probably easily reproduce it now.

**Comment by arpad-m** [comment]
Hmm yeah I really think that #9717 is the culprit, at least for the new issue. it's touched the code that is now printing:

```
2024-11-21T02:24:17.018130Z  INFO configurator_main_loop:reconfigure:apply_spec_sql: Applying phase CreateSuperUser
2024-11-21T02:24:17.020654Z  INFO configurator_main_loop:reconfigure:apply_spec_sql: Starting phase CreateSuperUser
2024-11-21T02:24:17.020665Z  INFO configurator_main_loop:reconfigure:apply_spec_sql:db_apply_changes{phase=CreateSuperUser}: Processing phase CreateSuperUser
2024-11-21T02:24:17.020675Z  INFO configurator_main_loop:reconfigure:apply_spec_sql:db_apply_changes{phase=CreateSuperUser}: Applying phase CreateSuperUser
PG:2024-11-21 02:24:17.007 GMT [4086451] LOG:  [NEON_SMGR] [shard 0] libpagestore: connected to 'postgresql://no_user@localhost:29434' with protocol version 2
PG:2024-11-21 02:24:17.029 GMT [4086451] LOG:  ConsoleURL not set, skipping forwarding
PG:2024-11-21 02:24:17.029 GMT [4086451] STATEMENT:  ​               DO $$​                       DECLARE​                     r text;​                         roles text[] := NULL;​                           dbs text[] := NULL;​                 BEGIN​                       IF NOT EXISTS (​                             SELECT FROM pg_catalog.pg_roles WHERE rolna>
2024-11-21T02:24:20.511160Z  INFO OpenTelemetry error: error sending request for url (http://localhost:4318/v1/traces)
PG:2024-11-21 02:24:47.034 GMT [4092035] LOG:  [WP] connecting with node 127.0.0.1:29437
PG:2024-11-21 02:24:47.034 GMT [4092035] LOG:  [WP] connecting with node 127.0.0.1:29443
PG:2024-11-21 02:24:47.034 GMT [4092035] LOG:  [WP] connecting with node 127.0.0.1:29446
PG:2024-11-21 02:24:47.034 GMT [4092035] LOG:  [WP] connected with node 127.0.0.1:29437
PG:2024-11-21 02:24:47.034 GMT [4092035] LOG:  [WP] connected with node 127.0.0.1:29443
PG:2024-11-21 02:24:47.034 GMT [4092035] LOG:  [WP] connected with node 127.0.0.1:29446
PG:2024-11-21 02:24:47.034 GMT [4092036] LOG:  [NEON_SMGR] [shard 0] libpagestore: connected to 'postgresql://no_user@localhost:29434' with protocol version 2
PG:2024-11-21 02:24:47.035 GMT [4092035] LOG:  [WP] received AcceptorGreeting from safekeeper 127.0.0.1:29443, term=2
PG:2024-11-21 02:24:47.035 GMT [4092035] LOG:  [WP] received AcceptorGreeting from safekeeper 127.0.0.1:29437, term=2
PG:2024-11-21 02:24:47.035 GMT [4092035] LOG:  [WP] proposer connected to quorum (2) safekeepers, propTerm=3
PG:2024-11-21 02:24:47.035 GMT [4092035] LOG:  [WP] requesting vote from 127.0.0.1:29437 for term 3
PG:2024-11-21 02:24:47.035 GMT [4092035] LOG:  [WP] requesting vote from 127.0.0.1:29443 for term 3
PG:2024-11-21 02:24:47.035 GMT [4092035] LOG:  [WP] received AcceptorGreeting from safekeeper 127.0.0.1:29446, term=2
PG:2024-11-21 02:24:47.035 GMT [4092035] LOG:  [WP] requesting vote from 127.0.0.1:29446 for term 3
PG:2024-11-21 02:24:47.035 GMT [4092035] LOG:  [WP] got VoteResponse from acceptor 127.0.0.1:29437, voteGiven=1, epoch=2, flushLsn=0/229C6A8, truncateLsn=0/1BB9638, timelineStartLsn=0/149EF10
PG:2024-11-21 02:24:47.035 GMT [4092035] LOG:  [WP] got VoteResponse from acceptor 127.0.0.1:29443, voteGiven=1, epoch=2, flushLsn=0/229C6A8, truncateLsn=0/1BB9638, timelineStartLsn=0/149EF10
PG:2024-11-21 02:24:47.035 GMT [4092035] LOG:  [WP] got votes from majority (2) of nodes, term 3, epochStartLsn 0/229C6A8, donor 127.0.0.1:29437, truncate_lsn 0/1BB9638
PG:2024-11-21 02:24:47.035 GMT [4092035] LOG:  [WP] sending elected msg to node 1 term=3, startStreamingAt=0/229C6A8 (lastCommonTerm=2), termHistory.n_entries=2 to 127.0.0.1:29437, timelineStartLsn=0/149EF10
PG:2024-11-21 02:24:47.036 GMT [4092035] LOG:  [WP] sending elected msg to node 3 term=3, startStreamingAt=0/229C6A8 (lastCommonTerm=2), termHistory.n_entries=2 to 127.0.0.1:29443, timelineStartLsn=0/149EF10
PG:2024-11-21 02:24:47.036 GMT [4092035] LOG:  [WP] WAL proposer starts streaming at 0/229C6A8
PG:2024-11-21 02:24:47.036 GMT [4092035] LOG:  [WP] got VoteResponse from acceptor 127.0.0.1:29446, voteGiven=1, epoch=2, flushLsn=0/229C6A8, truncateLsn=0/1BB9638, timelineStartLsn=0/149EF10
PG:2024-11-21 02:24:47.036 GMT [4092035] LOG:  [WP] sending elected msg to node 4 term=3, startStreamingAt=0/229C6A8 (lastCommonTerm=2), termHistory.n_entries=2 to 127.0.0.1:29446, timelineStartLsn=0/149EF10
2024-11-21T02:24:47.036488Z  INFO configurator_main_loop:reconfigure:apply_spec_sql:db_apply_changes{phase=CreateSuperUser}: successfully executed
```

you see the 30 second gap is somewhere in the `CreateSuperUser` phase the PR added. Note that the old code was also applying superuser privileges.

**Comment by arpad-m** [comment]
FTR it's not hard to make it reproduce, If I change it to run 16 copies of the test, I can run all 16 in parallel and there is always one or two that fails.

```diff
--- a/test_runner/regress/test_wal_acceptor.py
+++ b/test_runner/regress/test_wal_acceptor.py
@@ -1870,8 +1870,10 @@ def test_delete_timeline_under_load(neon_env_builder: NeonEnvBuilder):
 # Basic pull_timeline test.
 # When live_sk_change is False, compute is restarted to change set of
 # safekeepers; otherwise it is live reload.
-@pytest.mark.parametrize("live_sk_change", [False, True])
-def test_pull_timeline(neon_env_builder: NeonEnvBuilder, live_sk_change: bool):
+@pytest.mark.parametrize("num", range(16))
+def test_pull_timeline(neon_env_builder: NeonEnvBuilder, num: int):
+    live_sk_change = True
+    _num = num
     neon_env_builder.auth_enabled = True
```

**Comment by tristan957** [comment]
I'm working today and the first half of tomorrow. My understanding after talking with both @arpad-m and @MMeent is that Arpad is going to continue investigating from the storage side. Arpad, please let me know if that is wrong.

**Comment by jcsp** [comment]
Arseny to look into this once https://github.com/neondatabase/neon/pull/9915 is reviewed, to give a general sense of whether this area of code looks problematic

**Comment by arssher** [comment]
Tried to reproduce locally without success so far, looks like need to debug it on CI, unfortunately.

**Comment by arssher** [comment]
I can reproduce it locally only when running 32 or better 64 test instances in parallel (e.g using snippet above), causing CPU starvation. Sequentially reconfigure always takes about 1s as expected without https://github.com/neondatabase/neon/pull/9915 , and whole test takes about 6s. In CI test fails in about ~10% of runs. I don't see any particular misbehavior, so inclined to just increase the timeout -- probably CI runners are overloaded with current parallelism of 12 and 30s is not enough.

https://neondb.slack.com/archives/C059ZC138NR/p1733920416393659

**Comment by arssher** [comment]
Doesn't pop up in flaky tests dashboard after https://github.com/neondatabase/neon/pull/10088
